File: app/src/main/java/com/sevtinge/cemiuiler/module/securitycenter/LockOneHundredPoints.java
package com.sevtinge.cemiuiler.module.securitycenter;

import android.view.View;

import com.sevtinge.cemiuiler.module.base.BaseHook;
import com.sevtinge.cemiuiler.utils.PrefsUtils;
import de.robv.android.xposed.XC_MethodReplacement;
import de.robv.android.xposed.XposedBridge;
import io.luckypray.dexkit.DexKitBridge;
import io.luckypray.dexkit.builder.BatchFindArgs;
import io.luckypray.dexkit.descriptor.member.DexMethodDescriptor;
import io.luckypray.dexkit.enums.MatchType;

import java.lang.reflect.Method;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public class LockOneHundredPoints extends BaseHook {

    Class<?> mScoreManagerCls;
    Class<?> mMainContentFrameCls;

    @Override
    public void init() {

        mScoreManagerCls = findClassIfExists("com.miui.securityscan.scanner.ScoreManager");
        mMainContentFrameCls = findClassIfExists("com.miui.securityscan.ui.main.MainContentFrame");

        try {
            List<DexMethodDescriptor> result = Objects.requireNonNull(SecurityCenterDexKit.mSecurityCenterResultMap.get("ScoreManager"));
            for (DexMethodDescriptor descriptor : result) {
                Method lockOneHundredPoints = descriptor.getMethodInstance(lpparam.classLoader);
                log("lock 100 points method is "+ lockOneHundredPoints);
                if (lockOneHundredPoints.getReturnType() == int.class) {
                    XposedBridge.hookMethod(lockOneHundredPoints, XC_MethodReplacement.returnConstant(0));
                }
            }
        } catch (Throwable e) {
            e.printStackTrace();
        }

        /*findAndHookMethod(mScoreManagerCls, "B", new MethodHook() {
            @Override
            protected void before(MethodHookParam param) throws Throwable {
                if(PrefsUtils.mSharedPreferences.getBoolean("prefs_key_security_center_score", false)) param.setResult(0);
            }
        });*/

        findAndHookMethod(mMainContentFrameCls, "onClick", View.class, new MethodHook() {
            @Override
            protected void before(MethodHookParam param) throws Throwable {
                if(PrefsUtils.mSharedPreferences.getBoolean("prefs_key_security_center_score", false)) param.setResult(null);
            }
        });
    }
}

File: library/libhook/src/main/java/com/sevtinge/hyperceiler/libhook/rules/securitycenter/other/LockOneHundredPoints.kt
/*
  * This file is part of HyperCeiler.

  * HyperCeiler is free software: you can redistribute it and/or modify
  * it under the terms of the GNU Affero General Public License as
  * published by the Free Software Foundation, either version 3 of the
  * License.

  * This program is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU Affero General Public License for more details.

  * You should have received a copy of the GNU Affero General Public License
  * along with this program.  If not, see <https://www.gnu.org/licenses/>.

  * Copyright (C) 2023-2026 HyperCeiler Contributions
*/
package com.sevtinge.hyperceiler.libhook.rules.securitycenter.other

import android.view.View
import com.sevtinge.hyperceiler.libhook.base.BaseHook
import com.sevtinge.hyperceiler.libhook.utils.hookapi.dexkit.DexKit
import com.sevtinge.hyperceiler.libhook.utils.log.XposedLog
import io.github.kyuubiran.ezxhelper.core.finder.MethodFinder.`-Static`.methodFinder
import io.github.kyuubiran.ezxhelper.core.util.ClassUtil.loadClass
import io.github.kyuubiran.ezxhelper.xposed.dsl.HookFactory.`-Static`.createHook
import org.luckypray.dexkit.query.enums.StringMatchType
import java.lang.reflect.Method

object LockOneHundredPoints : BaseHook() {
    private val scoreOld by lazy<Method> {
        DexKit.findMember("LockOneHundredPoints1") {
            it.findClass {
                matcher {
                    className = "com.miui.securityscan.scanner.ScoreManager"
                }
            }.findMethod {
                matcher {
                    addUsingString("getMinusPredictScore", StringMatchType.Contains)
                    returnType = "int"
                }
            }.single()
        }
    }

    private val score by lazy<Method> {
        DexKit.findMember("LockOneHundredPoints2") {
            it.findClass {
                matcher {
                    className = "com.miui.securityscan.scanner.ScoreManager"
                }
            }.findMethod {
                matcher {
                    usingNumbers(41, 100, 0)
                    returnType = "int"
                }
            }.single()
        }
    }

    private val score3 by lazy<Method> {
        DexKit.findMember("LockOneHundredPointsField") {
            it.findClass {
                matcher {
                    className = "com.miui.securityscan.scanner.ScoreManager"
                }
            }.findMethod {
                matcher {
                    addUsingField {
                        type = "java.util.Map"
                    }
                    addInvoke {
                        addUsingField {
                            type = "int"
                        }
                        returnType = "int"
                    }
                }
            }.single()
        }
    }

    private val addRiskApp by lazy<Method> {
        DexKit.findMember("AddRiskApp") {
            it.findClass {
                matcher {
                    className = "com.miui.securityscan.scanner.ScoreManager"
                }
            }.findMethod {
                matcher {
                    addInvoke("Landroid/text/TextUtils;->isEmpty(Ljava/lang/CharSequence;)Z")
                    paramCount = 1
                    returnType = "void"
                }
            }.single()
        }
    }

    override fun init() {
        loadClass("com.miui.securityscan.ui.main.MainContentFrame").methodFinder()
            .filterByName("onClick")
            .filterByParamTypes(View::class.java)
            .first().createHook {
                returnConstant(null)
            }

        runCatching {
            scoreOld.createHook {
                returnConstant(0)
            }
            score.createHook {
                returnConstant(100)
            }
            score3.createHook {
                returnConstant(true)
            }
            addRiskApp.createHook {
                returnConstant(null)
            }
        }.onFailure {
            XposedLog.e(TAG, lpparam.packageName, "LockOneHundredPoints hook Failed: $it")
        }
    }
}
